bindings/rust/bench/src/harness.rs

```diff
@@ -12,14 +19,144 @@ pub enum Mode {
     Server,
 }
 
-pub trait TlsBenchHarness {
+// these parameters were the only ones readily usable for all three libaries:
+// s2n-tls, rustls, and openssl
+#[allow(non_camel_case_types)]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CipherSuite {
+    AES_128_GCM_SHA256,
+    AES_256_GCM_SHA384,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ECGroup {
+    SECP256R1,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct CryptoConfig {
+    pub cipher_suite: CipherSuite,
+    pub ec_group: ECGroup,
+}
+
+pub trait TlsBenchHarness: Sized {
+    /// Default harness
+    fn default() -> Result<Self, Box<dyn Error>> {
+        Self::new(&CryptoConfig {
+            cipher_suite: CipherSuite::AES_128_GCM_SHA256,
+            ec_group: ECGroup::SECP256R1,
+        })
+    }
+
     /// Initialize buffers, configs, and connections (pre-handshake)
-    fn new() -> Self;
+    fn new(crypto_config: &CryptoConfig) -> Result<Self, Box<dyn Error>>;
 
     /// Run handshake on initialized connection
-    /// Returns error if handshake already happened
-    fn handshake(&mut self) -> Result<(), &str>;
+    /// Returns error if handshake has already completed
+    fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
     /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
+
+    /// Get negotiated cipher suite
+    ///
+    /// Returns `Err` if handshake has not completed
+    fn get_negotiated_cipher_suite(&self) -> Result<CipherSuite, Box<dyn Error>>;
+
+    /// Get whether or negotiated version is TLS1.3
+    ///
+    /// Retuns `Err` if handshake has not completed
+    fn negotiated_tls13(&self) -> Result<bool, Box<dyn Error>>;
+}
+
+/// Wrapper of two shared buffers to pass as stream
+/// This wrapper `read()`s into one buffer and `write()`s to another
+#[derive(Clone)]
+pub struct ConnectedBuffer {
+    recv: Rc<RefCell<VecDeque<u8>>>,
+    send: Rc<RefCell<VecDeque<u8>>>,
+}
+
+impl ConnectedBuffer {
+    /// Make a new object with new internal buffers
+    pub fn new() -> Self {
+        ConnectedBuffer {
+            recv: Rc::new(RefCell::new(VecDeque::new())),
+            send: Rc::new(RefCell::new(VecDeque::new())),
+        }
+    }
+    /// Make a new object that shares internal buffers but swapped, ex.
+    /// `write()` writes to the buffer that the inverse `read()`s from
+    pub fn clone_inverse(&self) -> Self {
+        ConnectedBuffer {
+            recv: self.send.clone(),
+            send: self.recv.clone(),
+        }
+    }
+}
+
+impl Read for ConnectedBuffer {
+    fn read(&mut self, dest: &mut [u8]) -> Result<usize, std::io::Error> {
+        match self.recv.borrow_mut().read(dest) {
+            Ok(0) => Err(std::io::Error::new(ErrorKind::WouldBlock, "blocking")),
```

Comment:
Is it necessary to separately handle this case? It would be nice if we could just bubble up the result from the `borrow_mut().read(dest)` instead of having to break down the match statement.

If it is necessary to handle separately, we should add a comment explaining why.

Reply:
Rustls does expect a `WouldBlock` error, so we have to handle the 0 length case; I'll add a comment. It's probably because when using an actual TCP stream, there's a difference between reading in a message of length 0 and blocking (waiting for a response)?